### DISCUSSION TITLE
Adding more games to MichalOS

### DISCUSSION THREAD

**Discussion by shoaib-jamal** [prochazkaml/MichalOS#5] [Ideas]
Hello Michal, can you add some assembly projects in MichalOS. I have a list of them:
1. SorryAss - A F1 racing game ft. donkeys. A remake of Donkey.BAS originally by Bill Gates. [Get it Here](https://www.pouet.net/prod.php?which=67997)
2. Chrome's t-rex based bootsector game (512 bytes) written in 16-bit x86 assembly. [Get it Here](https://github.com/franeklubi/dino)
3. DosStars - draw animated stars on screen (can be used as a screen saver) [Get it Here](https://github.com/OrangeTide/dos-stars)
4. A recreation of flappy bird in asm [Get it Here](https://github.com/nanochess/fbird)
5. A bootable Pac-Man clone that fits inside the Master Boot Record [Get it here](https://github.com/guyhill/Boot-Man)
6. More ideas soon!

**Comment by prochazkaml** [comment]
Thanks for the ideas!

Pretty much any boot sector game should work under MichalOS with minor modifications (mainly returning back to the OS). Since they're so tiny, maybe I could compile a bunch of bootsector games into one program? 😏

**Reply by shoaib-jamal** [reply]
Yeah that's what I am saying dude 😌

**Reply by shoaib-jamal** [reply]
Hey I also have a pull request for ReadMe. Did you checked it?

**Reply by prochazkaml** [reply]
Yes, I did, and I made a couple of changes to it. Did _you_ check it?

**Reply by shoaib-jamal** [reply]
Ok 👍

**Reply by shoaib-jamal** [reply]
I have closed the pull request

**Comment by prochazkaml** [comment]
@shoaib-jamal Hey, I still keep this here for a reason! (When I eventually get back to working on MichalOS, I can easily access everything in here.)

**Reply by shoaib-jamal** [reply]
Oh, sure!

**Comment by shoaib-jamal** [comment]
@prochazkaml Hey, I was just wandering when I stumbled upon [this](https://github.com/Klaykap/BleskOS). BleskOS, it's nearly identical to MichalOS, written in assembly language and boasts an impressive array of functional resources, including USB, AHCI, Ethernet drivers, and more.  The developer is currently in the process of rewriting the entire OS in C. Considering these factors, I believe it would be worthwhile to dive into the current assembly code of BleskOS.

The license section says
> You can learn from BleskOS source code, but if you want to copy some part of it to your project, you have to let us know by email klaykap /@/ yandex /./ com

**Reply by prochazkaml** [reply]
Well, I'd say that the BleskOS developer is a very smart person for switching to C, as I'll probably end up doing the same (and, in fact, I have already done some 16-bit DOS programming in C using ia16-gcc, and I have to say, I'm impressed with the results).

While coding in assembly does bring in some benefits (such as the ability to hand-optimize your routines to such a degree which a typical compiler would not let you), in the end, I came to the conclusion that such benefits are outweighed by the fact that it is way easier to accidentally introduce dumb bugs in assembly (which, again, a compiler would not let you do - you wouldn't believe how many times I had to debug MichalOS just because of the fact that some program/kernel routine jumped into uninitialized memory, bringing the whole system down) and the overall tediousness of just working with it in such a large (OS) scale.

Also, BleskOS is a 32-bit OS. We're 16-bit over here, you simply cannot do things such as USB, AHCI or Ethernet, as they are simply out of reach for us. (With 16-bit, we don't have to write our own disk/gfx/keyboard drivers, as they're baked right into the BIOS, which we can use in 16-bit mode.)

**Reply by shoaib-jamal** [reply]
Ya, I really now think that C is a reliable option... 
Wish you all the best for rewriting MichalOS in C!

**Reply by prochazkaml** [reply]
Thank you! I'm sure I'll do it at some point, but I'm currently in the middle of another project, which coincidentally _is_ another operating system written in C:

![image](https://github.com/prochazkaml/MichalOS/assets/41787099/1188c28a-5b0f-45be-9c6a-dec82b84a8a8)

I mean, it's an embedded OS targeted at specific hardware, but still, it's a funny coincidence.

(that said, when I actually get around to working on MichalOS, I'll finally be able to put [my 486 laptop](https://github.com/prochazkaml/MichalOS/issues/13#issuecomment-1407490766) to good use! :D)

Also, with C, I can use ready-made libraries for some complex things, such as the wonderful [FatFs](http://elm-chan.org/fsw/ff/00index_e.html) library for handling even exFAT volumes! (I'm actually using this exact library in my current project, it's been a life-saver.)

**Reply by shoaib-jamal** [reply]
`*removed`

**Reply by prochazkaml** [reply]
Oh, sorry, but this is [onefetch](https://github.com/o2sh/onefetch) running on my host (Arch Linux) system. It's a tool for displaying info about the current git repository, where my project currently lives. 😅

In summary, it's the OS for my future watch (successor to [this](https://github.com/prochazkaml/tinyWatch)), but I won't get into the hardware details yet (as I'm still not decided on a few things).

All of the OS is stored in the microcontroller's flash memory, so a filesystem is not even necessary. Of course, it'll have an SD card for running external applications - otherwise it wouldn't really be an OS, would it?

**Comment by shoaib-jamal** [comment]
Oh, my mistake, so sorry, I didn't got a closer look 😅
Well I now understand it's an embedded watch OS, project looks cool anyway. If it can run external applications via SD card then it's really a nice start ~ plus, C gives it additional benefit - well it's really an OS even if is stored in the micro controller's flash memory.
Best of Luck!
